# Task
Create and run a quiz with 5-10 questions, track the user's score, display the final score and percentage, and save the results to a file named 'quiz_results.txt'.

## Define Quiz Questions and Answers

### Subtask:
Define a set of 5-10 questions and their corresponding correct answers to be used in the quiz.


**Reasoning**:
The subtask is to define quiz questions and answers. I will create a Python code block that generates a list of dictionaries, where each dictionary represents a question, a list of options, and the correct answer, as per the instructions.



In [1]:
quiz_questions = [
    {
        "question": "What is the capital of France?",
        "options": ["London", "Paris", "Berlin", "Rome"],
        "answer": "Paris"
    },
    {
        "question": "Which planet is known as the Red Planet?",
        "options": ["Earth", "Mars", "Jupiter", "Venus"],
        "answer": "Mars"
    },
    {
        "question": "What is 2 + 2?",
        "options": ["3", "4", "5", "6"],
        "answer": "4"
    },
    {
        "question": "Who wrote 'Romeo and Juliet'?",
        "options": ["Charles Dickens", "William Shakespeare", "Jane Austen", "Mark Twain"],
        "answer": "William Shakespeare"
    },
    {
        "question": "What is the largest ocean on Earth?",
        "options": ["Atlantic Ocean", "Indian Ocean", "Arctic Ocean", "Pacific Ocean"],
        "answer": "Pacific Ocean"
    },
    {
        "question": "What is the chemical symbol for water?",
        "options": ["O2", "H2O", "CO2", "NACL"],
        "answer": "H2O"
    },
    {
        "question": "Which country is famous for the Great Wall?",
        "options": ["India", "Japan", "China", "Egypt"],
        "answer": "China"
    }
]

print(f"Defined {len(quiz_questions)} quiz questions.")

Defined 7 quiz questions.


## Run Quiz and Track Score

### Subtask:
Present each question to the user, collect their answers, and track the number of correct responses.


**Reasoning**:
The subtask requires presenting questions, collecting answers, and tracking the score. I will write Python code to initialize a score, loop through the `quiz_questions` list, display each question and its options, get user input, and check if the answer is correct.



In [2]:
score = 0

for i, q_data in enumerate(quiz_questions):
    print(f"\nQuestion {i + 1}: {q_data['question']}")
    for j, option in enumerate(q_data['options']):
        print(f"{j + 1}. {option}")

    while True:
        try:
            user_input = input("Enter your answer (number): ")
            user_choice_index = int(user_input) - 1
            if 0 <= user_choice_index < len(q_data['options']):
                user_answer = q_data['options'][user_choice_index]
                break
            else:
                print("Invalid option number. Please try again.")
        except ValueError:
            print("Invalid input. Please enter a number.")

    if user_answer == q_data['answer']:
        print("Correct!")
        score += 1
    else:
        print(f"Wrong! The correct answer was: {q_data['answer']}")

print(f"\nQuiz finished! Your score: {score}")


Question 1: What is the capital of France?
1. London
2. Paris
3. Berlin
4. Rome
Enter your answer (number): Paris
Invalid input. Please enter a number.
Enter your answer (number): Rome
Invalid input. Please enter a number.
Enter your answer (number): London
Invalid input. Please enter a number.
Enter your answer (number): Berlin
Invalid input. Please enter a number.
Enter your answer (number): 2
Correct!

Question 2: Which planet is known as the Red Planet?
1. Earth
2. Mars
3. Jupiter
4. Venus
Enter your answer (number): 4
Wrong! The correct answer was: Mars

Question 3: What is 2 + 2?
1. 3
2. 4
3. 5
4. 6
Enter your answer (number): 2
Correct!

Question 4: Who wrote 'Romeo and Juliet'?
1. Charles Dickens
2. William Shakespeare
3. Jane Austen
4. Mark Twain
Enter your answer (number): 2
Correct!

Question 5: What is the largest ocean on Earth?
1. Atlantic Ocean
2. Indian Ocean
3. Arctic Ocean
4. Pacific Ocean
Enter your answer (number): 4
Correct!

Question 6: What is the chemical symbo

## Calculate and Display Results

### Subtask:
Calculate the final score and percentage of correct answers, then display them to the user.


**Reasoning**:
The subtask requires calculating and displaying the final score and percentage. I will write a Python code block to get the total number of questions, compute the percentage, and then print both the score and the formatted percentage.



In [3]:
total_questions = len(quiz_questions)
percentage = (score / total_questions) * 100

print(f"\nFinal Score: {score} out of {total_questions}")
print(f"Percentage: {percentage:.2f}%")


Final Score: 6 out of 7
Percentage: 85.71%


## Save Results to File

### Subtask:
Save the final score and percentage to a text file named 'quiz_results.txt'.


**Reasoning**:
The subtask requires saving the final score and percentage to a file. I will create a Python code block that opens a file named 'quiz_results.txt' in write mode, writes the formatted score and percentage, and ensures the file is closed properly using a 'with' statement.



In [4]:
output_filename = 'quiz_results.txt'

with open(output_filename, 'w') as f:
    f.write("--- Quiz Results ---\n")
    f.write(f"Final Score: {score} out of {total_questions}\n")
    f.write(f"Percentage: {percentage:.2f}%\n")
    f.write("--------------------\n")

print(f"Quiz results saved to '{output_filename}'")

Quiz results saved to 'quiz_results.txt'


## Final Task

### Subtask:
Confirm the successful creation and execution of the quiz program, including saving the results.


## Summary:

### Data Analysis Key Findings
*   A quiz program was successfully created and executed, featuring 7 distinct questions, which falls within the specified range of 5-10 questions.
*   The quiz interactively presented questions and options, collected user input, and provided immediate feedback on correctness.
*   Robust input validation was implemented to ensure users entered valid numeric choices for answers.
*   The user's final score was tracked, resulting in an example score of 6 out of 7 questions answered correctly.
*   The final percentage of correct answers was calculated and displayed as 85.71%.
*   All quiz results, including the final score (6 out of 7) and percentage (85.71%), were successfully saved to a text file named 'quiz_results.txt', with clear formatting.

### Insights or Next Steps
*   The implemented quiz system is functional and robust for basic use. Future enhancements could include adding a timer for each question or allowing users to review incorrect answers.
*   To make the quiz more engaging, consider implementing different question types (e.g., true/false, fill-in-the-blank) or varying difficulty levels.
